File: Algorithms/0318_Maximum_Product_of_Word_Lengths/Java/MaximumProductofWordLengthsSolution1.java

// Space: O(n)
// Time: O(n^2)

class Solution {
    public int maxProduct(String[] words) {
        int res = 0;
        for (int i = 0; i < words.length - 1; i++) {
            Set<Character> temp = new HashSet<>();
            for (char c : words[i].toCharArray()) {
                temp.add(c);
            }
            for (int j = i + 1; j < words.length; j++) {
                boolean isContain = false;
                for (char c : words[j].toCharArray()) {
                    if (temp.contains(c)) {
                        isContain = true;
                        break;
                    }
                }
                if (!isContain) res = Math.max(res, words[i].length() * words[j].length());
            }
        }
        return res;
    }
}



       

File: Algorithms/0318_Maximum_Product_of_Word_Lengths/Python/Maximum_Product_of_Word_Lengths_Solution_1.py

# Space: O(n)
# Time: O(n^2)

class Solution:
    def maxProduct(self, words):
        res = 0
        for i in range(len(words) - 1):
            for j in range(i + 1, len(words)):
                if len(set(words[i]).intersection(set(words[j]))) == 0:
                    res = max(res, len(words[i]) * len(words[j]))
        return res




